libs/validation/src/lib/DiscriminatedUnionValidator.spec.ts

```diff
@@ -0,0 +1,190 @@
+
+            expect(attemptToRegisterValidationFunctions).not.toThrow();
+        });
+    });
+
+    describe(`validate`, () => {
+        const allDiscriminantAndValidationFunctionPairs: [string, SimpleValidationFunction][] = [
+            ['circle', validateCircle],
+            ['square', validateSquare],
+            ['rectangle', validateRectangle],
+        ];
+
+        const discriminatedUnionValidator = new DiscriminatedUnionValidator(
+            myDiscriminants,
+            myDiscriminatorPropertyName
+        ).registerAllValidationFunctions(allDiscriminantAndValidationFunctionPairs);
+
+        describe(`when the input is valid`, () => {
+            const validCircles = [1, 3.14, 22.4].map((r) => new Circle(r));
+
+            const validSquares = [2, 99.9, 1200].map((l) => new Square(l));
+
+            const validRectangles = [
+                [2, 2],
+                [1.4, 5, 5],
+                [9, 80.1],
+            ].map(([h, w]) => new Rectangle(h, w));
+
+            it(`should return no errors`, () => {
+                [...validCircles, ...validSquares, ...validRectangles].forEach((validValue) => {
+                    const validationResult = discriminatedUnionValidator.validate(validValue);
+
+                    expect(validationResult).toEqual([]);
+                });
+            });
+        });
+
+        describe(`when the input is invalid`, () => {
+            describe(`when the input can be matched to its validation function by its discriminant`, () => {
+                const invalidValues = [new Circle(-5), new Square(-2), new Rectangle(-4.4, -23)];
+
+                invalidValues.forEach((invalidValue) => {
+                    const validationResult = discriminatedUnionValidator.validate(invalidValue);
+
+                    expect(validationResult.length).toBeGreaterThan(0);
+                });
+            });
+
+            describe(`when the input cannot be matched to its validation function by its discriminant`, () => {
+                describe(`when the input is null`, () => {
+                    const validationResult = discriminatedUnionValidator.validate(null);
+
+                    it(`should return an error`, () => {
+                        expect(validationResult.length).toBeGreaterThan(0);
+                    });
+                });
+
+                describe(`when the input is undefined`, () => {
+                    const validationResult = discriminatedUnionValidator.validate(undefined);
+
+                    it(`should return an error`, () => {
+                        expect(validationResult.length).toBeGreaterThan(0);
+                    });
+                });
+
+                describe(`when the input's discriminant is not in the allowed list`, () => {
+                    const badInput = {
+                        type: 'parallelogram',
+                        length: 22,
+                    };
+
+                    it('should throw', () => {
```

Comment:
Do we really want to throw here?

Reply:
We actually look for an error in the `expect`. I have updated the description now.